File: python/tests/test_image_base.py
"""IPT1 fence test: shared ``Image`` base + ``ImageKind`` discriminator.

Asserts (1) MS/Raman/IR subclass the new ``Image`` base, (2) the
``kind`` / ``spectral_axis`` / ``spectral_axis_kind`` discriminators
are wired correctly, and (3) byte-identical .tio round-trips for all
three image kinds are preserved (the refactor invariant).
"""
from __future__ import annotations

import numpy as np

from ttio.enums import IRMode, ImageKind, SpectralAxisKind
from ttio.image import Image
from ttio.ir_image import IRImage
from ttio.ms_image import MSImage
from ttio.providers import open_provider
from ttio.raman_image import RamanImage


def _ms() -> MSImage:
    cube = np.arange(2 * 3 * 4, dtype=np.float64).reshape(3, 2, 4)  # (h,w,sp)
    return MSImage(
        width=2, height=3, spectral_points=4, intensity=cube,
        mz_axis=np.linspace(100, 110, 4), pixel_size_x=1.0,
        pixel_size_y=2.0, scan_pattern="raster", title="t",
        isa_investigation_id="inv1",
    )


def _raman() -> RamanImage:
    cube = np.arange(3 * 3 * 5, dtype=np.float64).reshape(3, 3, 5) * 0.5
    wn = np.array([1000.0, 1100.0, 1200.0, 1300.0, 1400.0], dtype=np.float64)
    return RamanImage(
        width=3, height=3, spectral_points=5, intensity=cube,
        wavenumbers=wn, pixel_size_x=10.0, pixel_size_y=11.0,
        scan_pattern="raster", excitation_wavelength_nm=785.0,
        laser_power_mw=50.0, title="r", isa_investigation_id="inv2",
    )


def _ir() -> IRImage:
    cube = np.arange(3 * 3 * 5, dtype=np.float64).reshape(3, 3, 5) * 0.5
    wn = np.array([1000.0, 1100.0, 1200.0, 1300.0, 1400.0], dtype=np.float64)
    return IRImage(
        width=3, height=3, spectral_points=5, intensity=cube,
        wavenumbers=wn, pixel_size_x=10.0, pixel_size_y=11.0,
        scan_pattern="raster", mode=IRMode.ABSORBANCE,
        resolution_cm_inv=4.0, title="i", isa_investigation_id="inv3",
    )


def test_subclass_of_image() -> None:
    assert isinstance(_ms(), Image)
    assert isinstance(_raman(), Image)
    assert isinstance(_ir(), Image)


def test_kind_and_axis() -> None:
    m = _ms()
    assert m.kind == ImageKind.MS
    assert m.spectral_axis is m.mz_axis
    assert m.spectral_axis_kind == SpectralAxisKind.MZ

    r = _raman()
    assert r.kind == ImageKind.RAMAN
    assert r.spectral_axis is r.wavenumbers
    assert r.spectral_axis_kind == SpectralAxisKind.WAVENUMBER

    i = _ir()
    assert i.kind == ImageKind.IR
    assert i.spectral_axis is i.wavenumbers
    assert i.spectral_axis_kind == SpectralAxisKind.WAVENUMBER


def _roundtrip(img, cls):
    """Write ``img`` to a study group then read it back via ``cls``."""
    import tempfile
    from pathlib import Path
    with tempfile.TemporaryDirectory() as d:
        out = str(Path(d) / "image.tio")
        with open_provider(out, provider="hdf5", mode="w") as sp:
            study = sp.root_group().create_group("study")
            img.write_to(study)
        with open_provider(out, provider="hdf5", mode="r") as sp:
            study = sp.root_group().open_group("study")
            return cls.read_from(study)


def _assert_common_equal(read, orig) -> None:
    assert read is not None
    assert np.array_equal(read.intensity, orig.intensity)
    assert read.width == orig.width
    assert read.height == orig.height
    assert read.spectral_points == orig.spectral_points
    assert read.pixel_size_x == orig.pixel_size_x
    assert read.pixel_size_y == orig.pixel_size_y
    assert read.scan_pattern == orig.scan_pattern
    assert read.tile_size == orig.tile_size


def test_roundtrip_ms() -> None:
    orig = _ms()
    read = _roundtrip(orig, MSImage)
    _assert_common_equal(read, orig)
    np.testing.assert_array_equal(read.mz_axis, orig.mz_axis)
    assert read.spectral_axis_kind == SpectralAxisKind.MZ


def test_roundtrip_raman() -> None:
    orig = _raman()
    read = _roundtrip(orig, RamanImage)
    _assert_common_equal(read, orig)
    np.testing.assert_array_equal(read.wavenumbers, orig.wavenumbers)
    assert read.excitation_wavelength_nm == orig.excitation_wavelength_nm
    assert read.laser_power_mw == orig.laser_power_mw
    assert read.spectral_axis_kind == SpectralAxisKind.WAVENUMBER


def test_roundtrip_ir() -> None:
    orig = _ir()
    read = _roundtrip(orig, IRImage)
    _assert_common_equal(read, orig)
    np.testing.assert_array_equal(read.wavenumbers, orig.wavenumbers)
    assert read.mode == orig.mode
    assert read.resolution_cm_inv == orig.resolution_cm_inv
    assert read.spectral_axis_kind == SpectralAxisKind.WAVENUMBER

File: objc/Tests/TestImageBase.m
/*
 * TestImageBase.m
 *
 * OIT1 fence test for the TTIOImage base-class extraction. Verifies:
 *  - byte-identical .tio round-trips for MS / Raman / IR images
 *    (the common props + the distinct per-subclass axis/fields),
 *  - each subclass is a kind of TTIOImage,
 *  - the polymorphic kind / spectralAxis / spectralAxisKind accessors.
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 * Copyright (c) 2026 The Thalion Initiative
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Image/TTIOImage.h"
#import "Image/TTIOMSImage.h"
#import "Image/TTIORamanImage.h"
#import "Image/TTIOIRImage.h"
#import "ValueClasses/TTIOEnums.h"
#import "HDF5/TTIOHDF5Errors.h"
#import <unistd.h>

static NSString *baseImgPath(NSString *suffix)
{
    return [NSString stringWithFormat:@"/tmp/ttio_imgbase_%d_%@.tio",
            (int)getpid(), suffix];
}

static NSData *makeCube(NSUInteger w, NSUInteger h, NSUInteger sp)
{
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * sp * sizeof(double)];
    double *p = cube.mutableBytes;
    for (NSUInteger y = 0; y < h; y++)
        for (NSUInteger x = 0; x < w; x++)
            for (NSUInteger s = 0; s < sp; s++)
                p[(y * w + x) * sp + s] = (double)(x * 1000 + y * 13) + (double)s * 0.001;
    return cube;
}

static NSData *makeAxis(NSUInteger sp)
{
    NSMutableData *axis = [NSMutableData dataWithLength:sp * sizeof(double)];
    double *p = axis.mutableBytes;
    for (NSUInteger s = 0; s < sp; s++) p[s] = 100.0 + (double)s * 2.5;
    return axis;
}

void testImageBase(void)
{
    const NSUInteger W = 8, H = 6, SP = 4, TS = 32;
    NSData *cube = makeCube(W, H, SP);
    NSData *axis = makeAxis(SP);
    NSError *err = nil;

    // ---- MS ----
    {
        TTIOMSImage *img = [[TTIOMSImage alloc]
            initWithTitle:@"ms-title"
       isaInvestigationId:@"isa-ms"
          identifications:@[]
          quantifications:@[]
        provenanceRecords:@[]
                    width:W height:H spectralPoints:SP tileSize:TS
               pixelSizeX:1.5 pixelSizeY:2.5
              scanPattern:@"raster"
                     cube:cube mzAxis:axis];

        PASS([img isKindOfClass:[TTIOImage class]], "MSImage is a TTIOImage");
        PASS(img.kind == TTIOImageKindMS, "MSImage.kind == MS");
        PASS([img.spectralAxis isEqualToData:axis], "MSImage.spectralAxis == mzAxis");
        PASS(img.spectralAxisKind == TTIOSpectralAxisKindMZ, "MSImage.spectralAxisKind == MZ");

        NSString *path = baseImgPath(@"ms");
        unlink([path fileSystemRepresentation]);
        PASS([img writeToFilePath:path error:&err], "MS writes .tio");
        TTIOMSImage *back = [TTIOMSImage readFromFilePath:path error:&err];
        PASS(back != nil, "MS reads back");
        PASS([back.title isEqualToString:@"ms-title"], "MS title round-trips");
        PASS([back.isaInvestigationId isEqualToString:@"isa-ms"], "MS isaId round-trips");
        PASS(back.width == W && back.height == H && back.spectralPoints == SP,
             "MS dims round-trip");
        PASS(back.tileSize == TS, "MS tileSize round-trips");
        PASS(back.pixelSizeX == 1.5 && back.pixelSizeY == 2.5, "MS pixel sizes round-trip");
        PASS([back.scanPattern isEqualToString:@"raster"], "MS scanPattern round-trips");
        PASS([back.cube isEqualToData:cube], "MS cube bytes identical");
        PASS([back.mzAxis isEqualToData:axis], "MS mzAxis bytes identical");
        PASS([back.spectralAxis isEqualToData:axis], "MS spectralAxis bytes identical");
        PASS(back.kind == TTIOImageKindMS, "MS round-trip kind == MS");
        unlink([path fileSystemRepresentation]);
    }

    // ---- Raman ----
    {
        TTIORamanImage *img = [[TTIORamanImage alloc]
            initWithTitle:@"raman-title"
       isaInvestigationId:@"isa-raman"
          identifications:@[]
          quantifications:@[]
        provenanceRecords:@[]
                    width:W height:H spectralPoints:SP tileSize:TS
               pixelSizeX:3.0 pixelSizeY:4.0
              scanPattern:@"snake"
   excitationWavelengthNm:785.0 laserPowerMw:10.0
                     cube:cube wavenumbers:axis];

        PASS([img isKindOfClass:[TTIOImage class]], "RamanImage is a TTIOImage");
        PASS(img.kind == TTIOImageKindRaman, "RamanImage.kind == Raman");
        PASS([img.spectralAxis isEqualToData:axis], "RamanImage.spectralAxis == wavenumbers");
        PASS(img.spectralAxisKind == TTIOSpectralAxisKindWavenumber,
             "RamanImage.spectralAxisKind == Wavenumber");

        NSString *path = baseImgPath(@"raman");
        unlink([path fileSystemRepresentation]);
        PASS([img writeToFilePath:path error:&err], "Raman writes .tio");
        TTIORamanImage *back = [TTIORamanImage readFromFilePath:path error:&err];
        PASS(back != nil, "Raman reads back");
        PASS([back.title isEqualToString:@"raman-title"], "Raman title round-trips");
        PASS([back.isaInvestigationId isEqualToString:@"isa-raman"], "Raman isaId round-trips");
        PASS(back.width == W && back.height == H && back.spectralPoints == SP,
             "Raman dims round-trip");
        PASS(back.tileSize == TS, "Raman tileSize round-trips");
        PASS(back.pixelSizeX == 3.0 && back.pixelSizeY == 4.0, "Raman pixel sizes round-trip");
        PASS([back.scanPattern isEqualToString:@"snake"], "Raman scanPattern round-trips");
        PASS(back.excitationWavelengthNm == 785.0, "Raman excitation round-trips");
        PASS(back.laserPowerMw == 10.0, "Raman laser power round-trips");
        PASS([back.cube isEqualToData:cube], "Raman cube bytes identical");
        PASS([back.wavenumbers isEqualToData:axis], "Raman wavenumbers bytes identical");
        PASS([back.spectralAxis isEqualToData:axis], "Raman spectralAxis bytes identical");
        PASS(back.kind == TTIOImageKindRaman, "Raman round-trip kind == Raman");
        unlink([path fileSystemRepresentation]);
    }

    // ---- IR ----
    {
        TTIOIRImage *img = [[TTIOIRImage alloc]
            initWithTitle:@"ir-title"
       isaInvestigationId:@"isa-ir"
          identifications:@[]
          quantifications:@[]
        provenanceRecords:@[]
                    width:W height:H spectralPoints:SP tileSize:TS
               pixelSizeX:5.0 pixelSizeY:6.0
              scanPattern:@"raster"
                     mode:TTIOIRModeAbsorbance
          resolutionCmInv:4.0
                     cube:cube wavenumbers:axis];

        PASS([img isKindOfClass:[TTIOImage class]], "IRImage is a TTIOImage");
        PASS(img.kind == TTIOImageKindIR, "IRImage.kind == IR");
        PASS([img.spectralAxis isEqualToData:axis], "IRImage.spectralAxis == wavenumbers");
        PASS(img.spectralAxisKind == TTIOSpectralAxisKindWavenumber,
             "IRImage.spectralAxisKind == Wavenumber");

        NSString *path = baseImgPath(@"ir");
        unlink([path fileSystemRepresentation]);
        PASS([img writeToFilePath:path error:&err], "IR writes .tio");
        TTIOIRImage *back = [TTIOIRImage readFromFilePath:path error:&err];
        PASS(back != nil, "IR reads back");
        PASS([back.title isEqualToString:@"ir-title"], "IR title round-trips");
        PASS([back.isaInvestigationId isEqualToString:@"isa-ir"], "IR isaId round-trips");
        PASS(back.width == W && back.height == H && back.spectralPoints == SP,
             "IR dims round-trip");
        PASS(back.tileSize == TS, "IR tileSize round-trips");
        PASS(back.pixelSizeX == 5.0 && back.pixelSizeY == 6.0, "IR pixel sizes round-trip");
        PASS([back.scanPattern isEqualToString:@"raster"], "IR scanPattern round-trips");
        PASS(back.mode == TTIOIRModeAbsorbance, "IR mode round-trips");
        PASS(back.resolutionCmInv == 4.0, "IR resolution round-trips");
        PASS([back.cube isEqualToData:cube], "IR cube bytes identical");
        PASS([back.wavenumbers isEqualToData:axis], "IR wavenumbers bytes identical");
        PASS([back.spectralAxis isEqualToData:axis], "IR spectralAxis bytes identical");
        PASS(back.kind == TTIOImageKindIR, "IR round-trip kind == IR");
        unlink([path fileSystemRepresentation]);
    }
}
